<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Описание-проекта" data-toc-modified-id="Описание-проекта-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Описание проекта</a></span><ul class="toc-item"><li><span><a href="#Заказчик" data-toc-modified-id="Заказчик-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Заказчик</a></span></li><li><span><a href="#Задача" data-toc-modified-id="Задача-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Задача</a></span></li><li><span><a href="#Входные-данные" data-toc-modified-id="Входные-данные-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Входные данные</a></span><ul class="toc-item"><li><span><a href="#Описание-ресурсов" data-toc-modified-id="Описание-ресурсов-1.3.1"><span class="toc-item-num">1.3.1&nbsp;&nbsp;</span>Описание ресурсов</a></span><ul class="toc-item"><li><span><a href="#Перевозчики-(автопарки)" data-toc-modified-id="Перевозчики-(автопарки)-1.3.1.1"><span class="toc-item-num">1.3.1.1&nbsp;&nbsp;</span>Перевозчики (автопарки)</a></span></li><li><span><a href="#Экипажи" data-toc-modified-id="Экипажи-1.3.1.2"><span class="toc-item-num">1.3.1.2&nbsp;&nbsp;</span>Экипажи</a></span></li><li><span><a href="#Исполнители" data-toc-modified-id="Исполнители-1.3.1.3"><span class="toc-item-num">1.3.1.3&nbsp;&nbsp;</span>Исполнители</a></span></li><li><span><a href="#Машины" data-toc-modified-id="Машины-1.3.1.4"><span class="toc-item-num">1.3.1.4&nbsp;&nbsp;</span>Машины</a></span></li><li><span><a href="#Тарифы" data-toc-modified-id="Тарифы-1.3.1.5"><span class="toc-item-num">1.3.1.5&nbsp;&nbsp;</span>Тарифы</a></span></li><li><span><a href="#Драфты" data-toc-modified-id="Драфты-1.3.1.6"><span class="toc-item-num">1.3.1.6&nbsp;&nbsp;</span>Драфты</a></span></li><li><span><a href="#Заказы" data-toc-modified-id="Заказы-1.3.1.7"><span class="toc-item-num">1.3.1.7&nbsp;&nbsp;</span>Заказы</a></span></li><li><span><a href="#Карты" data-toc-modified-id="Карты-1.3.1.8"><span class="toc-item-num">1.3.1.8&nbsp;&nbsp;</span>Карты</a></span></li><li><span><a href="#Оплаты" data-toc-modified-id="Оплаты-1.3.1.9"><span class="toc-item-num">1.3.1.9&nbsp;&nbsp;</span>Оплаты</a></span></li><li><span><a href="#Другое" data-toc-modified-id="Другое-1.3.1.10"><span class="toc-item-num">1.3.1.10&nbsp;&nbsp;</span>Другое</a></span></li></ul></li></ul></li><li><span><a href="#План-работы" data-toc-modified-id="План-работы-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>План работы</a></span></li><li><span><a href="#Инструменты" data-toc-modified-id="Инструменты-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Инструменты</a></span><ul class="toc-item"><li><span><a href="#Импорт-библиотек" data-toc-modified-id="Импорт-библиотек-1.5.1"><span class="toc-item-num">1.5.1&nbsp;&nbsp;</span>Импорт библиотек</a></span></li><li><span><a href="#Определения-констант" data-toc-modified-id="Определения-констант-1.5.2"><span class="toc-item-num">1.5.2&nbsp;&nbsp;</span>Определения констант</a></span></li><li><span><a href="#Определения-функций" data-toc-modified-id="Определения-функций-1.5.3"><span class="toc-item-num">1.5.3&nbsp;&nbsp;</span>Определения функций</a></span><ul class="toc-item"><li><span><a href="#Загрузка-и-обзор-данных" data-toc-modified-id="Загрузка-и-обзор-данных-1.5.3.1"><span class="toc-item-num">1.5.3.1&nbsp;&nbsp;</span>Загрузка и обзор данных</a></span></li><li><span><a href="#Исправление-данных" data-toc-modified-id="Исправление-данных-1.5.3.2"><span class="toc-item-num">1.5.3.2&nbsp;&nbsp;</span>Исправление данных</a></span></li><li><span><a href="#Отображение-данных" data-toc-modified-id="Отображение-данных-1.5.3.3"><span class="toc-item-num">1.5.3.3&nbsp;&nbsp;</span>Отображение данных</a></span></li></ul></li></ul></li></ul></li><li><span><a href="#Сбор,-обзор-и-исправление-данных" data-toc-modified-id="Сбор,-обзор-и-исправление-данных-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Сбор, обзор и исправление данных</a></span><ul class="toc-item"><li><span><a href="#Итоги-обзора-и-исправления-данных" data-toc-modified-id="Итоги-обзора-и-исправления-данных-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Итоги обзора и исправления данных</a></span></li></ul></li><li><span><a href="#Расходы-каждого-пользователя-в-месяц" data-toc-modified-id="Расходы-каждого-пользователя-в-месяц-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Расходы каждого пользователя в месяц</a></span><ul class="toc-item"><li><span><a href="#Итоги-подготовки-информации-о-расходах:" data-toc-modified-id="Итоги-подготовки-информации-о-расходах:-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Итоги подготовки информации о расходах:</a></span></li></ul></li><li><span><a href="#Анализ-расходов" data-toc-modified-id="Анализ-расходов-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Анализ расходов</a></span><ul class="toc-item"><li><span><a href="#Итоги-анализа-расходов" data-toc-modified-id="Итоги-анализа-расходов-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Итоги анализа расходов</a></span></li></ul></li><li><span><a href="#Гипотезы-и-проверки" data-toc-modified-id="Гипотезы-и-проверки-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Гипотезы и проверки</a></span><ul class="toc-item"><li><span><a href="#Итоги-проверки-гипотез" data-toc-modified-id="Итоги-проверки-гипотез-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>Итоги проверки гипотез</a></span></li></ul></li><li><span><a href="#Общие-итоги" data-toc-modified-id="Общие-итоги-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Общие итоги</a></span><ul class="toc-item"><li><span><a href="#Рекомендации" data-toc-modified-id="Рекомендации-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>Рекомендации</a></span></li></ul></li></ul></div>

# Название проекта

Автор: Александр Лагутин.

Начат: 12 сентября 2022 г в 19:11.
Завершён: 1 июня 2022 г в 10:45.

Время подготовки: 12 рабочих часов (2 рабочих дня).

## Описание проекта

### Заказчик

Интернет-сервис грузоперевозок MOVER.

Основная деятельность: организация перевозок отдельных (не сборных) грузов.

Идея: подбор исполнителей для разовой грузоперевозки. Один исполнитель - один заказчик (нет сборных грузов, и одновременных заказов у одного исполнителя). Перевозки внутри Москвы, фургонами до 1,5 т.

### Задача

Выявить масштабы бизнеса:
- обороты,
- неуплаты,
- количество заказчиков,
- водителей.

Обзор парка водителей:
- активность;
- характер исполнения заказов;
- характеристики машин.

Далее будут поставлены другие задачи.

### Входные данные

Доступ к API.

#### Описание ресурсов

    Правило:
    ЧТО_ДЕЛАЕТ: /кто/который/над_кем/которым?как=так

    support - администратор
    freighter - автопарк
    employer - исполнитель
    customer - заказчик

    me
    {id}
    all

    crews
    transports
    drafts
    orders

##### Перевозчики (автопарки)

    Все:
    GET: /support/me/freighters.json

##### Экипажи

    Все:
    GET: /freighter/{freighter_id}/crews.json
    CRUD:
    POST: /freighter/{freighter_id}/crews.json
    GET: /freighter/{freighter_id}/crews/{crew_id}.json
    POST: /freighter/{freighter_id}/crews/{crew_id}.json
    DELETE: /freighter/{freighter_id}/crews/{crew_id}.json

##### Исполнители

    Все:
    GET: /freighter/{freighter_id}/employers.json
    CRUD:
    POST: /freighter/{freighter_id}/employers.json
    GET: /freighter/{freighter_id}/employers/{employer_id}.json
    POST: /freighter/{freighter_id}/employers/{employer_id}.json
    ?

##### Машины

    Все:
    GET: /freighter/{freighter_id}/transports.json
    CRUD:
    POST: /freighter/{freighter_id}/transports.json
    GET: /freighter/{freighter_id}/transports/{transport_id}.json
    POST: /freighter/{freighter_id}/transports/{transport_id}.json
    ?
    Все типы машин:
    GET: /transport/models.json
    Все свойства машин:
    GET: /features.json

##### Тарифы

    Все:
    GET: /tariff.json
    Доступны в данной точке:
    GET: /tariff/tier.json?lat={latitude}&lon={longitude}
    Сколько машин свободны возле точки:
    GET: /tariff/tier/counts.json?lat={latitude}&lon={longitude}

##### Драфты

    Создать драфт:
    POST: /support/me/orders/drafts.json
    Драфты, прошедшие поиск:
    GET: /support/me/drafts/searched.json
    GET: /support/me/drafts/{draft_id}.json

##### Заказы

    Все:
    GET: /support/me/orders.json
    GET: /freighter/all/orders.json
    Уведомления по заказу:
    GET: /freighter/me/orders/{order_id}/notifications.json
    Изменить статус заказа:
    PUT: /freighter/{freighter_id}/orders/{order_id}/status.json?status={current_status}

##### Карты

    Адрес по координате:
    GET: /geo/location.json?location=lat,lon
    Город по IP:
    GET: /geo/ip.json
    Координаты по адресу, и подсказки адресов:
    GET: /geo/address.json?address={адрес или часть адреса}
    Построить маршрут:
    GET: /geo/route.json?avoidTtk={true|false}location[]={latitude,longitude}&location[]={latitude,longitude}&....

##### Оплаты

    История счетов:
    /me/invoices.json
    Статистика перевозок по компаниям:
    GET: /freighter/{freighter_id}/stat/orders.json

##### Другое

    Получить количество грузчиков:
    GET: /employer/me/crew/loaders.json
    Получить статус исполнителя:
    GET: /employer/me/status.json
    Изменить статус исполнителя:
    PUT: /employer/me/status.json?status={current_status}

### План работы

Действия:
- [ ] Импортировать библиотеки, описать функции, которые потребуются для расчёта и диаграмм LTV, ROI, RR и CR.
- [ ] Загрузить и изучить таблицы.
- [ ] Подготовить данные:
  - [ ] привести к норме названия столбцов;
  - [ ] привести данные к нужным типам;
  - [ ] обработать пропуски:
    - [ ] выбрать, чем заполнить (или оставить) и почему;
    - [ ] осуществить заполение, если требуется;
    - [ ] описать причины, которые могли привести к пропускам.
- [ ] Анализировать обобщённые данные:
  - [ ] Создать профили пользователей.
    - [ ] Найти минимальную и даты привлечения пользователей.
  - [ ] Построить таблицы "Количество пользователей" - "Доля плативших", сгруппированные по странам, устройствам и источникам.
  - [ ] Описать выводы.
- [ ] Описать общий вывод:
  - [ ] причины неэффективности;
  - [ ] рекомендации отделу маркетинга.

### Инструменты

Используем [Anaconda](https://www.anaconda.com/), окружение `ds_da_practicum_env.yml` "для Mac OS" (от 2022-03-23).

#### Импорт библиотек

In [1]:
%matplotlib inline

import datetime
import re

import pandas as pd
import numpy as np
import scipy.stats as st
import matplotlib.pyplot as plt
import seaborn as sns

import mover

#### Определения констант

In [2]:
FIGSIZE = (12, 6)
mover = mover.Session('79060597475', '2022', logfile='mover_log.txt')

#### Определения функций

##### Загрузка и обзор данных

In [3]:
def load(tablename, ext='csv'):
    """Load a table from file
    """
    filename = tablename + '.' + ext
    try:
        df = pd.read_csv(filename)
    except FileNotFoundError:
        df = pd.read_csv('https://code.s3.yandex.net/datasets/' + filename)
    correct_column_names(df)  #  переводит в snake_case
    survey(df)  # даёт обзор возможных проблем с таблицей
    return df

In [4]:
def survey(data):
    """Display summaries to detect data problems
    """
    display(data.sample(5))
    print(f'Полностью совпадающих строк: {data.duplicated().sum()}\n')
    data.info()
    display(get_nans(data))
    print(get_filled_rows_share(data))

In [5]:
def get_nans(data):
    nans = []
    for column in data.columns:
        missing = data[column].isna().sum()  # Подсчет количества отсутствующих значений
        part = round(missing * 100 / len(data), 2) # Подсчет доли отсутствующих значений
        uniques = data[column].sort_values().nunique()
        nans.append([column, missing, part, uniques])

    return pd.DataFrame(
        data=nans, columns=['column', 'na_count', 'na_%', 'unique_count']
    ).sort_values(['na_count', 'unique_count'], ascending=False).set_index('column')

In [6]:
def get_filled_rows_share(df):
    share = 1 - df.isna().any(axis=1).sum() / len(df)
    return f'Заполненных строк без пропусков: {share:.02%}'

##### Исправление данных

In [7]:
date_pattern = (
    '^(?:\\d{4})-(?:\\d{2})-(?:\\d{2})'  # date
    '[ T](?:\\d{2}):(?:\\d{2}):(?:\\d{2}(?:\\.\\d*)?)'  # time
    '(?:(?:-(?:\\d{2}):(?:\\d{2})|Z)?)$'  # timezone
)

print(re.match(date_pattern, '2021-11-04 11:11:11') is not None)

# visits['session_start'] = pd.to_datetime(visits['session_start'])
# visits['session_end'] = pd.to_datetime(visits['session_start'])

True


In [8]:
def correct_column_names(df):
    df.columns = df.columns.str.strip()
    df.columns = df.columns.apply(to_snake_case)
    # df.columns = df.columns.str.lower()
    # df.columns = df.columns.str.replace(' ', '_')

def to_snake_case(name):
    name = re.sub('(.)([A-Z][a-z]+)', r'\1_\2', name)
    name = re.sub('__([A-Z])', r'_\1', name)
    name = re.sub('([a-z0-9])([A-Z])', r'\1_\2', name)
    return name.lower()

In [9]:
def fillnaby(data, column, groupby, func='median'):
    data[column] = data[column].fillna(
        data.groupby(groupby)[column].transform(func))

##### Отображение данных

In [10]:
def headline(line, width=50):
    """Возвращает заголовок, набранный "простым текстом".
    """
    return f'\n{line.upper():=^{width}}'

In [11]:
# можно было бы рисовать `bar` в случаях, когда вариантов мало.
# однако по неясной причине `bar` не рисуется, и не вызывает ошибки,
# а вызывает лишь зацикливание при построении первого же графика

def survey_columns(df):
    for column in nans_before.index:
        print(headline(column))

        uniques = realty[column].sort_values().unique()
        if len(uniques) < 20:
            print(uniques)
        print(f'Data type: {str(realty[column].dtype):>10}')
        print(f'Unique values: {len(uniques):6}')

        missing = realty[column].isna().sum()
        if missing:
            print(f'Not avaliable: {nans_before.loc[column, "na_part"]:.02%}')
        else:
            print('All values is avaliable.')

        try:
            realty.plot(
                kind='hist',
                y=column,
               bins=50,
            )
            plt.show()
        except Exception as e:
            print(str(e).capitalize())

In [12]:
# формирует и выводит график, подстроенный под наши нужды
def histf(data, column, show=True, bins=100, qtl=1, start=None,
          stop=None, figsize=FIGSIZE, **kwargs):
    fixed = {
        'y': column,
        'kind': 'hist',
        'bins': min(bins, len(data[column].unique())),
        'figsize': figsize,
        'title': column,
        'legend': False,
        'ec': 'black',
    }
    if 'range' not in kwargs.keys():
        if not start:
            start = data[column].min()
        if not stop:
            stop = data[column].quantile(qtl)
        kwargs['range'] = (start, stop)
    data.plot(**fixed, **kwargs)
    if show:
        plt.show()

def histf2(data, column1, column2, **kwargs):
    histf(column1, data=data, show=False, **kwargs)
    histf(column2, data=data, show=False, **kwargs)
    plt.show()

# два графика: все случаи, и те, что менее указанного квантиля
def dhistf(column, data, qtl=0.975, figsize=FIGSIZE, **kwargs):
    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=figsize)
    fig.suptitle('This is a somewhat long figure title', fontsize=16)
    
    ax1.set_title('full_data')
    ax2.set_title(f'{qtl:.2%} quantile')
    histf(column, data=data, ax=ax1, subplots=True, show=False, **kwargs)
    histf(column, data=data, ax=ax2, subplots=True, show=False, qtl=qtl, **kwargs)
    plt.show()
# Развитая форма этого чудовища должна принимать список квантилей,
# и строить столько графиков, сколько квантилей указано,
# и качестве центра диапазона выбирать медиану.
# Пора перейти к аналитике.

In [13]:
def typical(data, column, ratio=1.5):
    q1 = data[column].quantile(0.25)
    q3 = data[column].quantile(0.75)
    iqr = q3 - q1
    whiskers = iqr * ratio
    low = q1 - whiskers
    high = q3 + whiskers
    
    print(f'{column.title()}: min = {data[column].min()},'
          f' max = {data[column].max()},'
          f' normal range = [{low:.2f} ... {high:.2f}]')

    return data.query(f'@low <= {column} <= @high')

In [14]:
def find_cat_and_num_cols(df):
    # find categorical and numerical columns
    
    categorical_columns = [c for c in df.columns if df[c].dtype.name == 'object']
    numerical_columns   = [c for c in df.columns if df[c].dtype.name != 'object']
    
    return categorical_columns, numerical_columns

categorical_properties, quantitative_properties = find_cat_and_num_cols(pd.DataFrame(data=[]))

key_column = ''

def batch_corr(data, column=key_column):
    print(headline('Коэффициенты линейной корреляции'))
    display(
            data
            .reindex(columns=(categorical_properties + quantitative_properties))
            .corr()
            [column].sort_values(ascending=False)
    )

def batch_scatter(data, column=key_column):
    for prop in quantitative_properties:
        if prop == column:
            continue
        data.plot(
            kind='scatter', alpha=0.2, figsize=FIGSIZE,
            y=column,
            x=prop,
        )
    plt.show()

def batch_box(data, column=key_column, by_props=categorical_properties):
    for prop in by_props:
        data.boxplot(
            column=column,
            by=prop,
            figsize=FIGSIZE,
        )
    plt.show()

## Сбор, обзор и исправление данных

In [18]:
parks = mover.get('/support/me/freighters.json')
print(parks.keys())
for p in parks['freighters']:
    print(f"{p['id']:3} : {p['name']}")

dict_keys(['success', 'freighters'])
122 : 1
  5 : Gazelchik
 14 : LiveCargo
  1 : MOVER
  8 : MOVER - СОЦ
 10 : MOVER - СТАР
110 : MOVER СПБ
  2 : Mover Упаковка
  6 : Pereezdik
  4 : Podorojnik
  3 : Timegruz
  7 : ааааа
  9 : АВАНГАРД ПЛЮС
111 : Автофлот
 88 : Бирюков Сергей Сергеевич ИП
 42 : ГепардСити
106 : Грузовичкоф
143 : Грузовичкоф XXL
136 : Грузовичкоф Казань
146 : Грузовичкоф Краснодар
120 : Грузовичкоф Псков
108 : Грузовичкоф Спб
 11 : Евростар
105 : ЕвроТранс
124 : ИП Абдряхимов Алексей Ярулович
 27 : ИП Абышев Эрнис Абдыбекович
 86 : ИП Алдонов Александр Александрович
 43 : ИП Асланов Анвар Джамалович
 65 : ИП Астахова Юлия Владимировна
 64 : ИП Бабин Евгений Евгеньевич
 89 : ИП Балов Виталий Федорович
 17 : ИП Баранов Алексей Николаевич
 63 : ИП Башурин Евгений Владимирович
133 : ИП Безпалов Игорь Георгиевич
 49 : ИП Белкин Евгений Иванович
 40 : ИП Брагуца Александр Тимофеевич
 37 : ИП Брус Дмитрий Геннадьевич
 26 : ИП Бурова Татьяна Владимировна
 95 : ИП Бурынкин Дми

### Итоги обзора и исправления данных

- Первый;
- Второй;
- Третий.

## Расходы каждого пользователя в месяц

### Итоги подготовки информации о расходах:

- Первый;
- Второй;
- Третий.

## Анализ расходов

### Итоги анализа расходов

- Первый;
- Второй;
- Третий.

## Гипотезы и проверки

### Итоги проверки гипотез

- Первый;
- Второй;
- Третий.

## Общие итоги

- Исследовано:
  - выборка пяти сотен пользователей из разных регионов, подключившихся в 2018 году;
  - полная статистика их потребления звонков, коротких сообщений и интернет-трафика.
- **Подтверждённые выводы:**
  - Пользователь тарифа **"Ultra"** приносит **в среднем на 33% больше выручки**, чем пользователь тарифа "Smart".
  - **Средняя выручка** с пользователя **в Москве не отличается** от средней выручки с пользователя в других регионах.
- Также обнаружено, что:
  - Пользователей тарифа "Smart" в 4 раза больше;
  - Среди тех, кто потребляет много, популярней тариф "Ultra";
  - В среднем пользователи "Smart" потребляют столько, сколько включено в базовый пакет;
  - При относительно небольшом увеличении потребления их чек заметно вырастает;
  - Если потребление превышает 100 звонков (или 750 минут), или 110 сообщений, или 25 ГБ в месяц - пользователи предпочитают тариф "Ultra";
  - Пользователи "Ultra" потребляют только треть от базового пакета;
  - И 85% из них платят только базовый платёж.
- Предположения:
  - Пользователи "Smart":
    - сознательно ограничивают своё потребление;
    - при повышении потребления переходят на гораздо более содержательный "Ultra".
  - Пользователи "Ultra":
    - при повышении потребления вдвое не станут платить вдвое больше.

### Рекомендации

**Описание**

- пишем от лица "мы", употребляем подходящие времена
- Пиши краткие промежуточные выводы.
- Сделай общий вывод по итогам проекта в отдельном разделе.
- Сохрани идеи и код, который помогает принимать промежуточные решения.
- удали неработающий код
- комментарии ставь перед кодом
- вынеси комментарии в маркдаун
- подробно описывай процесс, что и почему ты делаешь
- за каждым пояснением - ячейка с кодом. Одна ячейка - один шаг
- перед отправкой - перезапусти вессь проект, убедись, что все ячейки сработали верно

**Вывод данных**
- Каждому объекту данных (фрейм, график) - дай название, подпись, подпиши столбцы и оси.
- Таблицы давай выборками, .head() и .tail()
- таблицы выводи по display()  # from IPython.display import display

**Итоги**
- краткий обзор проведённой работы: что выполнили, рассчитали:
- главные выводы: ответы на поставленные вопросы
- рекомендации: как избежать неправильных данных, как отобрать признаки для прогнозирования в будущем

На что обращают внимание при проверке проекта:
- [ ] Как вы описываете выявленные в данных проблемы?
- [ ] Как готовите датасет к анализу?
- [ ] Какие строите графики для распределений и как их объясняете?
- [ ] Как рассчитываете стандартное отклонение и дисперсию?
- [ ] Формулируете ли альтернативную и нулевую гипотезы?
- [ ] Какие методы применяете, чтобы их проверить?
- [ ] Объясняете результат проверки гипотезы или нет?
- [ ] Соблюдаете ли структуру проекта и поддерживаете аккуратность кода?
- [ ] Какие выводы делаете?
- [ ] Оставляете ли комментарии к шагам?